File: backend/python-api/app/repository/token_repo.py
from __future__ import annotations

from datetime import datetime, timezone

from sqlalchemy import select, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.domain.token import (
    EmailVerificationToken,
    PasswordResetToken,
    RefreshToken,
)


class TokenRepository:
    def __init__(self, session: AsyncSession):
        self._session = session

    # --- Refresh tokens ---

    async def create_refresh_token(self, token: RefreshToken) -> None:
        self._session.add(token)
        await self._session.flush()

    async def find_refresh_token_by_hash(self, token_hash: str) -> RefreshToken | None:
        stmt = select(RefreshToken).where(RefreshToken.token_hash == token_hash)
        result = await self._session.execute(stmt)
        return result.scalar_one_or_none()

    async def revoke_refresh_token_by_hash(self, token_hash: str) -> None:
        await self._session.execute(
            update(RefreshToken).where(RefreshToken.token_hash == token_hash).values(revoked=True)
        )
        await self._session.flush()

    async def revoke_token_family(self, family_id: str) -> None:
        await self._session.execute(
            update(RefreshToken).where(RefreshToken.family_id == family_id).values(revoked=True)
        )
        await self._session.flush()

    async def revoke_all_user_refresh_tokens(self, user_id: int) -> None:
        await self._session.execute(
            update(RefreshToken)
            .where(RefreshToken.user_id == user_id, RefreshToken.revoked.is_(False))
            .values(revoked=True)
        )
        await self._session.flush()

    # --- Password reset tokens ---

    async def create_password_reset_token(self, token: PasswordResetToken) -> None:
        self._session.add(token)
        await self._session.flush()

    async def find_password_reset_token(self, token_str: str) -> PasswordResetToken | None:
        stmt = select(PasswordResetToken).where(
            PasswordResetToken.token == token_str,
            PasswordResetToken.used.is_(False),
            PasswordResetToken.expires_at > datetime.now(timezone.utc),
        )
        result = await self._session.execute(stmt)
        return result.scalar_one_or_none()

    async def invalidate_password_reset_tokens(self, user_id: int) -> None:
        await self._session.execute(
            update(PasswordResetToken)
            .where(PasswordResetToken.user_id == user_id, PasswordResetToken.used.is_(False))
            .values(used=True)
        )
        await self._session.flush()

    async def mark_password_reset_token_used(self, token_id: int) -> None:
        await self._session.execute(
            update(PasswordResetToken).where(PasswordResetToken.id == token_id).values(used=True)
        )
        await self._session.flush()

    # --- Email verification tokens ---

    async def create_email_verification_token(self, token: EmailVerificationToken) -> None:
        self._session.add(token)
        await self._session.flush()

    async def find_email_verification_token(self, token_str: str) -> EmailVerificationToken | None:
        stmt = select(EmailVerificationToken).where(
            EmailVerificationToken.token == token_str,
            EmailVerificationToken.used.is_(False),
            EmailVerificationToken.expires_at > datetime.now(timezone.utc),
        )
        result = await self._session.execute(stmt)
        return result.scalar_one_or_none()

    async def mark_email_verification_token_used(self, token_id: int) -> None:
        await self._session.execute(
            update(EmailVerificationToken).where(EmailVerificationToken.id == token_id).values(used=True)
        )
        await self._session.flush()

File: templates/backends/go-gin/internal/repository/token_repo.go
package repository

import (
	"context"
	"errors"

	"saas-starter/backend/go-api/internal/domain"
	"saas-starter/backend/go-api/internal/platform/database"

	"gorm.io/gorm"
)

type tokenRepo struct {
	db *gorm.DB
}

func NewTokenRepository(db *gorm.DB) TokenRepository {
	return &tokenRepo{db: db}
}

func (r *tokenRepo) conn(ctx context.Context) *gorm.DB {
	return database.DBFromContext(ctx, r.db)
}

// --- Refresh Token (stored as SHA-256 hash) ---

func (r *tokenRepo) CreateRefreshToken(ctx context.Context, token *domain.RefreshToken) error {
	return r.conn(ctx).Create(token).Error
}

func (r *tokenRepo) FindRefreshTokenByHash(ctx context.Context, tokenHash string) (*domain.RefreshToken, error) {
	var token domain.RefreshToken
	if err := r.conn(ctx).Where("token_hash = ?", tokenHash).First(&token).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, domain.ErrNotFound
		}
		return nil, err
	}
	return &token, nil
}

func (r *tokenRepo) RevokeRefreshTokenByHash(ctx context.Context, tokenHash string) error {
	return r.conn(ctx).Model(&domain.RefreshToken{}).
		Where("token_hash = ?", tokenHash).
		Update("revoked", true).Error
}

func (r *tokenRepo) RevokeTokenFamily(ctx context.Context, familyID string) error {
	return r.conn(ctx).Model(&domain.RefreshToken{}).
		Where("family_id = ? AND revoked = ?", familyID, false).
		Update("revoked", true).Error
}

func (r *tokenRepo) RevokeAllUserRefreshTokens(ctx context.Context, userID uint) error {
	return r.conn(ctx).Model(&domain.RefreshToken{}).
		Where("user_id = ? AND revoked = ?", userID, false).
		Update("revoked", true).Error
}

// --- Password Reset Token ---

func (r *tokenRepo) CreatePasswordResetToken(ctx context.Context, token *domain.PasswordResetToken) error {
	return r.conn(ctx).Create(token).Error
}

func (r *tokenRepo) FindPasswordResetToken(ctx context.Context, tokenStr string) (*domain.PasswordResetToken, error) {
	var token domain.PasswordResetToken
	if err := r.conn(ctx).Where("token = ? AND used = ?", tokenStr, false).First(&token).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, domain.ErrNotFound
		}
		return nil, err
	}
	return &token, nil
}

func (r *tokenRepo) InvalidatePasswordResetTokens(ctx context.Context, userID uint) error {
	return r.conn(ctx).Model(&domain.PasswordResetToken{}).
		Where("user_id = ? AND used = ?", userID, false).
		Update("used", true).Error
}

func (r *tokenRepo) MarkPasswordResetTokenUsed(ctx context.Context, id uint) error {
	return r.conn(ctx).Model(&domain.PasswordResetToken{}).
		Where("id = ?", id).
		Update("used", true).Error
}

// --- Email Verification Token ---

func (r *tokenRepo) CreateEmailVerificationToken(ctx context.Context, token *domain.EmailVerificationToken) error {
	return r.conn(ctx).Create(token).Error
}

func (r *tokenRepo) FindEmailVerificationToken(ctx context.Context, tokenStr string) (*domain.EmailVerificationToken, error) {
	var token domain.EmailVerificationToken
	if err := r.conn(ctx).Where("token = ? AND used = ?", tokenStr, false).First(&token).Error; err != nil {
		if errors.Is(err, gorm.ErrRecordNotFound) {
			return nil, domain.ErrNotFound
		}
		return nil, err
	}
	return &token, nil
}

func (r *tokenRepo) MarkEmailVerificationTokenUsed(ctx context.Context, id uint) error {
	return r.conn(ctx).Model(&domain.EmailVerificationToken{}).
		Where("id = ?", id).
		Update("used", true).Error
}
